src/chainstate/stacks/index/node.rs

```diff
@@ -0,0 +1,2809 @@
+        let mut ptrs = [TriePtr::default(); 256];
+        for i in 0..48 {
+            let c = node48.ptrs[i].chr();
+            ptrs[c as usize] = node48.ptrs[i].clone();
+        }
+        TrieNode256 {
+            path: node48.path.clone(),
+            ptrs: ptrs,
+        }
+    }
+}
+
+impl TrieNode for TrieNode4 {
+    fn id(&self) -> u8 {
+        TrieNodeID::Node4
+    }
+
+    fn empty() -> TrieNode4 {
+        TrieNode4 {
+            path: vec![],
+            ptrs: [TriePtr::default(); 4],
+        }
+    }
+
+    fn walk(&self, chr: u8) -> Option<TriePtr> {
+        for i in 0..4 {
+            if self.ptrs[i].id() != TrieNodeID::Empty && self.ptrs[i].chr() == chr {
+                return Some(self.ptrs[i].clone());
+            }
+        }
+        return None;
+    }
+
+    fn to_bytes(&self, ret: &mut Vec<u8>) -> () {
+        let id = self.id();
+        ptrs_to_bytes(id, &self.ptrs, ret);
+        path_to_bytes(&self.path, ret);
+    }
+
+    fn to_consensus_bytes(&self, ret: &mut Vec<u8>) -> () {
+        let id = self.id();
+        ptrs_to_consensus_bytes(id, &self.ptrs, ret);
+        path_to_bytes(&self.path, ret);
+    }
+
+    fn byte_len(&self) -> usize {
+        get_ptrs_byte_len(&self.ptrs) + get_path_byte_len(&self.path)
+    }
+
+    fn from_bytes<R: Read>(r: &mut R) -> Result<TrieNode4, Error> {
+        let mut ptrs_slice = [TriePtr::default(); 4];
+        ptrs_from_bytes(TrieNodeID::Node4, r, &mut ptrs_slice)?;
+        let path = path_from_bytes(r)?;
+
+        Ok(TrieNode4 {
+            path,
+            ptrs: ptrs_slice,
+        })
+    }
+
+    fn insert(&mut self, ptr: &TriePtr) -> bool {
+        if self.replace(ptr) {
+            return true;
+        }
+
+        for i in 0..4 {
+            if self.ptrs[i].id() == TrieNodeID::Empty {
+                self.ptrs[i] = ptr.clone();
+                return true;
+            }
+        }
+        return false;
+    }
+
+    fn replace(&mut self, ptr: &TriePtr) -> bool {
+        for i in 0..4 {
+            if self.ptrs[i].id() != TrieNodeID::Empty && self.ptrs[i].chr() == ptr.chr() {
+                self.ptrs[i] = ptr.clone();
+                return true;
+            }
+        }
+        return false;
+    }
+
+    fn ptrs(&self) -> &[TriePtr] {
+        &self.ptrs
+    }
+
+    fn try_as_node4(&self) -> Option<TrieNodeType> { Some(TrieNodeType::Node4(self.clone())) }
+    fn try_as_node16(&self) -> Option<TrieNodeType> { Some(TrieNodeType::Node16(TrieNode16::from_node4(&self))) }
+    fn try_as_node48(&self) -> Option<TrieNodeType> { Some(TrieNodeType::Node48(TrieNode48::from_node16(&TrieNode16::from_node4(&self)))) }
+    fn try_as_node256(&self) -> Option<TrieNodeType> { Some(TrieNodeType::Node256(TrieNode256::from_node48(&TrieNode48::from_node16(&TrieNode16::from_node4(&self))))) }
+    fn try_as_leaf(&self) -> Option<TrieNodeType> { None }
+}
+
+impl TrieNode for TrieNode16 {
```

Comment:
In both cases, the `TriePtr` struct captures both the keys (`.chr`) and pointers (`.ptr`).

Unlike the paper, we aren't using SIMD instructions to search the set of `chr` values for a match -- this would (1) be unsafe, (2) be x86-specific, (3) not offer significant performance improvements, since in our case, the bulk of the time spent on inserts comes from the cryptographic hash function, and the bulk of the time spent on reads comes from disk seeks.

Reply:
One thing we _might_ do is store the `TriePtr`s in a `TrieNode16` in order by `chr`, and then do a binary search traversal on `insert()` and `replace()`.  However, it's unclear to me that even this would be worthwhile, since the pipeline stalls incurred by branching could end up being _slower_ than a sequential scan of the `TriePtr` slice (which most CPUs have highly optimized opcodes for doing -- scanning memory sequentially).